### PR TITLE
Raise minimum React version

### PR DESCRIPTION
**Pull request by connor-baer**
Addresses [DSYS-371](https://sumupteam.atlassian.net/browse/DSYS-371). Relates to #1998.

## Purpose

Circuit UI has supported [React 18](https://reactjs.org/blog/2022/03/29/react-v18.html) since #1554 while continuing to support older versions of React (>=16.8). React 18 introduced APIs to support the new concurrent rendering mode. Circuit UI must adopt these APIs, which requires raising the minimum React version to React 18.

## Approach and changes

- Raise the minimum version of the `react` and `react-dom` peer dependencies to >=18
- Deprecate the `uniqueId` util and replace it with the new [`useId` hook](https://beta.reactjs.org/reference/react/useId)

## Definition of done

* [x] Development completed
* [x] Reviewers assigned
* [x] Unit and integration tests
* [x] Meets minimum browser support
* [x] Meets accessibility requirements

### CODE REVIEW

**Comment by connor-baer** [comment]
Running into https://github.com/storybookjs/storybook/issues/21034.

File: .changeset/chilled-walls-learn.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@sumup/circuit-ui': major
+---
+
+Raised the minimum version of `react` and `react-dom` to >=18.
```
Comment (connor-baer) [inline .changeset/chilled-walls-learn.md:5]:
```suggestion
Raised the minimum version of the `react` and `react-dom` peer dependencies to >=18.
```

File: packages/circuit-ui/components/Carousel/__snapshots__/Carousel.spec.js.snap [hunk 2]
```diff
@@ -555,15 +555,15 @@ exports[`Carousel styles should render with children as a function 1`] = `
       class="circuit-42 circuit-43"
     >
       <span
-        aria-labelledby="progress-bar_6"
+        aria-labelledby=":r2:"
```
Comment (robinmetral) [inline packages/circuit-ui/components/Carousel/__snapshots__/Carousel.spec.js.snap:558]: oh interesting 👀 first time I see a snapshot of `useId`

File: packages/circuit-ui/components/Checkbox/Checkbox.tsx [hunk 3]
```diff
@@ -188,8 +187,9 @@ export const Checkbox = forwardRef(
     }: CheckboxProps,
     ref: CheckboxProps['ref'],
   ) => {
-    const id = customId || uniqueId('checkbox_');
-    const validationHintId = uniqueId('validation_hint-');
+    const id = useId();
+    const checkboxId = customId || id;
```
Comment (robinmetral) [inline packages/circuit-ui/components/Checkbox/Checkbox.tsx:191]: Could this be

```suggestion
    const id = customId || useId();
```

...so we don't generate one if developers pass it?

File: packages/circuit-ui/components/Input/__snapshots__/Input.spec.tsx.snap [hunk 4]
```diff
@@ -178,7 +178,7 @@ exports[`Input Styles should prioritize disabled over warning styles 1`] = `
   >
     <label
       class="circuit-1"
-      for="input_23"
+      for=":rm:"
```
Comment (robinmetral) [inline packages/circuit-ui/components/Input/__snapshots__/Input.spec.tsx.snap:181]: For me? 🤩

File: packages/circuit-ui/components/Popover/Popover.tsx [hunk 5]
```diff
@@ -285,8 +284,8 @@ export const Popover = ({
   const zIndex = useStackContext();
   const triggerKey = useRef<TriggerKey | null>(null);
   const menuEl = useRef<HTMLDivElement>(null);
-  const triggerId = useMemo(() => uniqueId('trigger_'), []);
-  const menuId = useMemo(() => uniqueId('popover_'), []);
+  const triggerId = useId();
+  const menuId = useId();
```
Comment (robinmetral) [inline packages/circuit-ui/components/Popover/Popover.tsx:288]: So nice! ✨

**Review by robinmetral — COMMENTED** [review]
Looks great!

File: packages/circuit-ui/components/Checkbox/Checkbox.tsx (hunk 3 above)
Comment (connor-baer) [reply packages/circuit-ui/components/Checkbox/Checkbox.tsx:191]: That would violate one of the [rules of hooks](https://legacy.reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level):

> Don’t call Hooks inside loops, conditions, or nested functions.

Comment (robinmetral) [reply packages/circuit-ui/components/Checkbox/Checkbox.tsx:191]: oopsies 😅 I thought I was missing something here

**Review by robinmetral — APPROVED** [review]
Well this looks good to me! I'm happy to take another look when this is marked ready for review if you'd like.

**Pull request merged**